sympy/core/relational.py

```diff
@@ -1131,3 +1033,269 @@ def _eval_relation(cls, lhs, rhs):
+    """
+    Resolves cmp(lhs, rhs) to S.true if cmp returns true.
+    or to S.false if cmp returns false
+    or to None if cmp returns none.
+    """
+    val = cmp(lhs, rhs, op)
+    if val is None:
+        return cls(lhs, rhs, evaluate=False)
+    elif isinstance(val,bool):
+        if val is True:
+            return S.true
+        else:
+            return S.false
+    else:
+        return val
+
+
+def cmp(lhs, rhs, op):
+    """
+    Returns Fuzzy-Valued (True, False,None) truth-values, of a lhs and a rhs.
+    None-means that lhs cannot be compared against lhs.
+
+    To Override the default comparison, implement _eval_is_X, except for Equality which should override
+    _eval_Eq, for legacy reasons
+    Assumes, lhs and rhs are sympified
+
+    Examples:
+    >>> from sympy.core.relational import cmp
+    >>> from sympy import sympify
+    >>> from sympy import Symbol
+    >>> x = Symbol('x')
+    >>> two = sympify(2)
+    >>> cmp(x,two, "<") is None
+
+    >>> from sympy.core.relational import cmp
+    >>> from sympy import sympify
+    >>> from sympy import Symbol
+    >>> x = Symbol('x')
+    >>> two = sympify(2)
+    >>> cmp(x,two, "=") is None
+
+    >>> from sympy.core.relational import cmp
+    >>> from sympy import sympify
+    >>> from sympy import Symbol
+    >>> three = sympify(3)
+    >>> two = sympify(2)
+    >>> cmp(three,two, "<") is False
+
+    >>> from sympy.core.relational import cmp
+    >>> from sympy import sympify
+    >>> from sympy import Symbol
+    >>> three = sympify(3)
+    >>> two = sympify(2)
+    >>> cmp(three,two, ">") is True
+
+    """
+    assert op in ("<", ">", "<=", ">=", "=",'!=')
+
+    # if one of the either the lhs, or the rhs has an _eval_is_X method
+    # return the result, otherwise do nothing
+    retval = None
+    eval_str = "_eval_"
+    if op == "<":
+      eval_str = eval_str + "is_lt"
+    elif op == ">":
+        eval_str = eval_str + "is_gt"
+    elif op == "<=":
+        eval_str = eval_str + "is_le"
+    elif op == ">=":
+        eval_str = eval_str + "is_ge"
+    elif op == "=" or op == '!=':
+        eval_str = eval_str + "Eq"
+
+    if hasattr(lhs, eval_str):
+        retval = getattr(lhs, eval_str)(rhs)
+    elif hasattr(rhs, eval_str):
+        retval = getattr(rhs, eval_str)(lhs)
+
+    if isinstance(retval, BooleanAtom):
+        if op == "!=":
+            return retval is not S.true
+        else:
+            return retval is S.true
+    elif retval is not None and op == "!=":
+        return not retval
+    elif retval is not None:
+        return retval
+
+    # if the operation is equality, then return the results of equality
+    if op == "=":
+        return _cmp_eq(lhs, rhs)
+
+    # if the operation is inequality, then return the results of equality, negated
+    elif op == '!=':
+        is_eq = _cmp_eq(lhs, rhs)
+        if is_eq is not None:
+            return not is_eq
+        else:
+            return is_eq
+
+    for me in (lhs, rhs):
```

Comment:
We might want to make this return False in the first case; for `nan`, maybe None instead of False.

Reply:
See discussion [here](https://groups.google.com/forum/#!topic/sympy/W0iKfAGsc2E).